x/concentrated-liquidity/position.go

```diff
@@ -621,17 +610,30 @@ func (k Keeper) RemovePositionIdToLock(ctx sdk.Context, positionId, underlyingLo
 	store.Delete(lockIdKey)
 }
 
-// positionHasUnderlyingLockInState checks if a given positionId has a corresponding lock in state.
-func (k Keeper) positionHasUnderlyingLockInState(ctx sdk.Context, positionId uint64) (bool, error) {
+// PositionHasActiveUnderlyingLockInState checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, removes the lock ID from the position ID to lock ID mapping and returns false.
+func (k Keeper) PositionHasActiveUnderlyingLockInState(ctx sdk.Context, positionId uint64) (bool, uint64, error) {
 	// Get the lock ID for the position.
-	_, err := k.GetLockIdFromPositionId(ctx, positionId)
-	if err == nil || !errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return true, nil
-	}
+	lockId, err := k.GetLockIdFromPositionId(ctx, positionId)
 	if errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return false, nil
+		return false, 0, nil
+	} else if err != nil {
+		return false, 0, err
+	}
+
+	// Check if the underlying lock is mature.
+	lockIsMature, err := k.isLockMature(ctx, lockId)
+	if err != nil {
+		return true, 0, err
+	}
+	if lockIsMature {
```

Comment:
I do not see a reason why we would want to retain the lock to position ID state entry, as in an ideal world it would immediately be removed via hook once it was mature

Reply:
(the only thing stopping me from implementing this via hook is import cycle between CL and lockup)

Reply:
IMO I think, since adding this to the lockup mature hook is not viable, this is the next best option (to call this every time we check a positions underlying lock). Please lmk what you think @mattverse @p0mvn

Reply:
The import cycle issue seems solvable with a minor package refactor, but it seems like triggering via hook automatically could expose a spam vector.

Lazily clearing these during checks (or some other regular user interaction) seems reasonable to me.

Reply:
> , but it seems like triggering via hook automatically could expose a spam vector.

I think I am missing something, how can using hooks expose use to spam vectors?

Reply:
@mattverse the problem with this though, is `RemovePositionIdToLockIfMature` would / should always get called with `PositionHasActiveUnderlyingLock`, and if future calls did not remember to call the `remove` portion then we would have unexpected behavior

Reply:
@czarcas7ic in that case, can we not make it the opposite way, where we make a method that encapsulates the logic for `removePosition`, lets say `RemovePositionIdToLockIfMature` and then within that function we can call `PositionHasActiveUnderlyingLock`. 

Apologies if I am being picky about method names here, I just think that method names that performs state changes should be explicit in the method name or else would also expose us to unexpected behaviors or bugs

Reply:
I'm supportive of having `RemovePositionIdToLockIfMature` as a separate method and either calling `RemovePositionIdToLockIfMature` and `PositionHasActiveUnderlyingLock` from another method that encapsulates them or simply calling them together.

The fact that for this use case we need to perform the 2 actions, doesn't justify encapsulating them under a method called `PositionHasActiveUnderlyingLock`. From the name, I would never be able to tell that it is a mutative method